### ISSUE TITLE
Scipy is not installed due to "Error: 'numpy' must be installed before running the build."

### ISSUE THREAD

**Issue by killangell** [kivy/python-for-android#2509]
### Checklist

- [ ] the issue is indeed a bug and not a support request
- [ ] issue doesn't already exist: https://github.com/kivy/python-for-android/issues
- [ ] I have a short, runnable example that reproduces the issue
- [ ] I reproduced the problem with the latest development version (`p4a.branch = develop`)
- [ ] I used the grave accent (aka backticks) to format code or logs when appropriated

### Versions

- Python: 3.8
- OS:
- Kivy:
- Cython:
- OpenJDK:

kivy@kivy-virtual-machine:~/work/project/biga_futures/package$ p4a --version
2021.09.05

### Description

Scipy is pending on numpy as I saw scipy recipe below:

    depends = ['setuptools', 'cython', 'numpy', 'lapack']

It's failed been installed for error below, more logs see Logs section.
"Error: 'numpy' must be installed before running the build."

### buildozer.spec

Command:
  p4a apk --private $HOME/work/project/biga_futures/  --package=org.kivy.biga_futures  --name "biga_futures" --version 0.1 --bootstrap=sdl2 --requirements=python3,kivy,pandas,typing_extensions,chardet,requests,urllib3,psutil,jwt,cryptography,numpy,scipy --permission PERMISSIONS INTERNET

No matter numpy is in the '--requirements' or not, nothing different.

Spec file:
```
// REPLACE ME: Paste your buildozer.spec file here
```

### Logs

[INFO]:    pandas said it is already built, skipping
[INFO]:    Building scipy for armeabi-v7a
[INFO]:    scipy apparently isn't already in site-packages
[INFO]:    Building compiled components in scipy
[INFO]:    -> directory context /home/kivy/.local/share/python-for-android/build/other_builds/scipy/armeabi-v7a__ndk_target_21/scipy
[INFO]:    -> running python3 setup.py build_ext -v -j 1
           working: Error: 'numpy' must be installed before running the build.                                                                                                                                                                                                Exception in thread background thread for pid 1206223:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/kivy/.local/lib/python3.8/site-packages/sh.py", line 1683, in wrap
    fn(*rgs, **kwargs)
  File "/home/kivy/.local/lib/python3.8/site-packages/sh.py", line 2662, in background_thread
    handle_exit_code(exit_code)
  File "/home/kivy/.local/lib/python3.8/site-packages/sh.py", line 2349, in fn
    return self.command.handle_command_exit_code(exit_code)
  File "/home/kivy/.local/lib/python3.8/site-packages/sh.py", line 905, in handle_command_exit_code
    raise exc
sh.ErrorReturnCode_1:

  RAN: /home/kivy/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build/python3 setup.py build_ext -v -j 1

  STDOUT:
Error: 'numpy' must be installed before running the build.

  STDERR:

Traceback (most recent call last):
  File "/home/kivy/.local/bin/p4a", line 8, in <module>
    sys.exit(main())
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/entrypoints.py", line 18, in main
    ToolchainCL()
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/toolchain.py", line 725, in __init__
    getattr(self, command)(args)
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/toolchain.py", line 153, in wrapper_func
    build_dist_from_args(ctx, dist, args)
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/toolchain.py", line 212, in build_dist_from_args
    build_recipes(build_order, python_modules, ctx,
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/build.py", line 580, in build_recipes
    recipe.build_arch(arch)
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/recipe.py", line 1011, in build_arch
    self.build_compiled_components(arch)
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/recipes/scipy/__init__.py", line 16, in build_compiled_components
    super().build_compiled_components(arch)
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/recipe.py", line 1022, in build_compiled_components
    shprint(hostpython, 'setup.py', self.build_cmd, '-v',
  File "/home/kivy/.local/lib/python3.8/site-packages/pythonforandroid/logger.py", line 167, in shprint
    for line in output:
  File "/home/kivy/.local/lib/python3.8/site-packages/sh.py", line 953, in next
    self.wait()
  File "/home/kivy/.local/lib/python3.8/site-packages/sh.py", line 879, in wait
    self.handle_command_exit_code(exit_code)
  File "/home/kivy/.local/lib/python3.8/site-packages/sh.py", line 905, in handle_command_exit_code
    raise exc
sh.ErrorReturnCode_1:

  RAN: /home/kivy/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build/python3 setup.py build_ext -v -j 1

  STDOUT:
Error: 'numpy' must be installed before running the build.

**Comment by killangell** [comment]
> 
> 
> try checking out p4a at [7f4652a](https://github.com/kivy/python-for-android/commit/7f4652afee5afe06ed811a40125f7cf6069728c8) and see if it works there.

Thanks for the help, with this patch, seems things got different than before.
I got new error : arm-linux-androideabi/bin/ld: error: cannot find -lunwind

While I searched in my ENV, it's should be there: 
/home/kivy/Android/android-ndk-r19b/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/arm-linux-androideabi/libunwind.a
/home/kivy/Android/android-ndk-r19b/sources/cxx-stl/llvm-libc++/libs/armeabi-v7a/libunwind.a

Any suggestion for how to move forward, appreciate for the help.

> error below
compile options: '-I/home/kivy/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/numpy/core/include -I/home/kivy/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build/Lib/site-packages/numpy/core/include -I/home/kivy/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/Include -I/home/kivy/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/native-build -c'
clang -L/home/kivy/.local/share/python-for-android/build/libs_collections/unnamed_dist_1/armeabi-v7a -L/home/kivy/.local/share/python-for-android/build/other_builds/python3-libbz2-liblzma/armeabi-v7a__ndk_target_21/python3/android-build -lpython3.8 -target armv7a-linux-androideabi21 -fomit-frame-pointer -march=armv7-a -mfloat-abi=softfp -mfpu=vfp -mthumb -fPIC -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions --sysroot=/home/kivy/Android/android-ndk-r19b/platforms/android-21/arch-arm -L/home/kivy/Android/android-ndk-r19b/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/arm-linux-androideabi/lib64 -L/home/kivy/.local/share/python-for-android/build/python-installs/unnamed_dist_1/numpy/core/lib -target armv7a-linux-androideabi21 -fomit-frame-pointer -march=armv7-a -mfloat-abi=softfp -mfpu=vfp -mthumb -fPIC -I/home/kivy/.local/share/python-for-android/build/other_builds/python3-libbz2-liblzma/armeabi-v7a__ndk_target_21/python3/Include -DANDROID -D__ANDROID_API__=21 -I/home/kivy/Android/android-ndk-r19b/sysroot/usr/include/arm-linux-androideabi -I/home/kivy/.local/share/python-for-android/build/python-installs/unnamed_dist_1/include/python3.8 --sysroot=/home/kivy/Android/android-ndk-r19b/platforms/android-21/arch-arm -I/home/kivy/Android/android-ndk-r19b/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1 -I/home/kivy/Android/android-ndk-r19b/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include build/temp.linux-x86_64-3.8/scipy/cluster/_vq.o -Lbuild/temp.linux-x86_64-3.8 -o build/lib.linux-x86_64-3.8/scipy/cluster/_vq.cpython-38-x86_64-linux-gnu.so -Wl,--version-script=build/temp.linux-x86_64-3.8/link-version-scipy.cluster._vq.map
/home/kivy/Android/android-ndk-r19b/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lunwind
/home/kivy/Android/android-ndk-r19b/toolchains/llvm/prebuilt/linux-x86_64/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld: error: cannot find -lunwind

**Comment by killangell** [comment]
I tried to copy threads.h as follows:
kivy@kivy-virtual-machine:/$ cp /home/kivy/work/android_source/prebuilts/ndk/current/platforms/android-18/arch-arm/usr/include/linux/threads.h /home/kivy/.local/share/python-for-android/build/other_builds/hostpython3/desktop/hostpython3/Include/

> Another issue appeared:

    ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:618:12: error: use of undeclared identifier '_active_cb_jac_in_dvode__user__routines'; did you mean 'get_active_cb_jac_in_dvode__user__routines'?
    return _active_cb_jac_in_dvode__user__routines;
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           get_active_cb_jac_in_dvode__user__routines
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:617:43: note: 'get_active_cb_jac_in_dvode__user__routines' declared here
static cb_jac_in_dvode__user__routines_t *get_active_cb_jac_in_dvode__user__routines(void) {
                                          ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:652:9: warning: unused variable 'ipar' [-Wunused-variable]
    int ipar=(*ipar_cb_capi);
        ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:648:9: warning: unused variable 'ml' [-Wunused-variable]
    int ml=(*ml_cb_capi);
        ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:649:9: warning: unused variable 'mu' [-Wunused-variable]
    int mu=(*mu_cb_capi);
        ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:651:12: warning: unused variable 'rpar' [-Wunused-variable]
    double rpar=(*rpar_cb_capi);
           ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:803:8: error: unknown type name 'thread_local'
static F2PY_THREAD_LOCAL_DECL cb_f_in_zvode__user__routines_t *_active_cb_f_in_zvode__user__routines = NULL;
       ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:89:32: note: expanded from macro 'F2PY_THREAD_LOCAL_DECL'
#define F2PY_THREAD_LOCAL_DECL thread_local
                               ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:803:62: error: expected ';' after top level declarator
static F2PY_THREAD_LOCAL_DECL cb_f_in_zvode__user__routines_t *_active_cb_f_in_zvode__user__routines = NULL;
                                                             ^
                                                             ;
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:806:45: error: use of undeclared identifier '_active_cb_f_in_zvode__user__routines'
    cb_f_in_zvode__user__routines_t *prev = _active_cb_f_in_zvode__user__routines;
                                            ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:807:5: error: use of undeclared identifier '_active_cb_f_in_zvode__user__routines'
    _active_cb_f_in_zvode__user__routines = ptr;
    ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:812:12: error: use of undeclared identifier '_active_cb_f_in_zvode__user__routines'; did you mean 'get_active_cb_f_in_zvode__user__routines'?
    return _active_cb_f_in_zvode__user__routines;
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           get_active_cb_f_in_zvode__user__routines
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:811:41: note: 'get_active_cb_f_in_zvode__user__routines' declared here
static cb_f_in_zvode__user__routines_t *get_active_cb_f_in_zvode__user__routines(void) {
                                        ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:843:9: warning: unused variable 'ipar' [-Wunused-variable]
    int ipar=(*ipar_cb_capi);
        ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:842:12: warning: unused variable 'rpar' [-Wunused-variable]
    double rpar=(*rpar_cb_capi);
           ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:995:8: error: unknown type name 'thread_local'
static F2PY_THREAD_LOCAL_DECL cb_jac_in_zvode__user__routines_t *_active_cb_jac_in_zvode__user__routines = NULL;
       ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:89:32: note: expanded from macro 'F2PY_THREAD_LOCAL_DECL'
#define F2PY_THREAD_LOCAL_DECL thread_local
                               ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:995:64: error: expected ';' after top level declarator
static F2PY_THREAD_LOCAL_DECL cb_jac_in_zvode__user__routines_t *_active_cb_jac_in_zvode__user__routines = NULL;
                                                               ^
                                                               ;
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:998:47: error: use of undeclared identifier '_active_cb_jac_in_zvode__user__routines'
    cb_jac_in_zvode__user__routines_t *prev = _active_cb_jac_in_zvode__user__routines;
                                              ^
build/src.linux-x86_64-3.8/scipy/integrate/vodemodule.c:999:5: error: use of undeclared identifier '_active_cb_jac_in_zvode__user__routines'
    _active_cb_jac_in_zvode__user__routines = ptr;
    ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
14 warnings and 20 errors generated.

**Comment by killangell** [comment]
Besides my last comment, I missed another error for building lapack, it cannot find cblas.h.

> Path on my ENV
kivy@kivy-virtual-machine:/$ sudo find / -name cblas.h
[sudo] password for kivy:
/home/kivy/.local/share/python-for-android/build/other_builds/lapack/armeabi-v7a__ndk_target_21/lapack/build/install/include/cblas.h
/home/kivy/.local/share/python-for-android/build/other_builds/lapack/armeabi-v7a__ndk_target_21/lapack/build/include/cblas.h
/home/kivy/.local/share/python-for-android/build/other_builds/lapack/armeabi-v7a__ndk_target_21/lapack/CBLAS/include/cblas.h
find: ‘/run/user/1000/gvfs’: Per

> Error
creating /tmp/tmps2il8dd4/tmp
creating /tmp/tmps2il8dd4/tmp/tmps2il8dd4
compile options: '-I/home/kivy/.local/share/python-for-android/build/other_builds/lapack/armeabi-v7a__ndk_target_21/lapack/build/install/lib -I -c'
ccache: /tmp/tmps2il8dd4/source.c
/tmp/tmps2il8dd4/source.c:1:10: fatal error: 'cblas.h' file not found
#include <cblas.h>
         ^~~~~~~~~
1 error generated.

**Comment by mzakharo** [comment]
As per comment [here](https://github.com/kivy/python-for-android/blob/develop/pythonforandroid/recipes/lapack/__init__.py#L2):  make sure you have latest cmake.  Patches were tested using  NDK r19c, and target API 28.

**Comment by killangell** [comment]
> 
> 
> As per comment [here](https://github.com/kivy/python-for-android/blob/develop/pythonforandroid/recipes/lapack/__init__.py#L2): make sure you have latest cmake. Patches were tested using NDK r19c, and target API 28.

I'm using r19b, not sure if that the difference is causing the problem on my side.
I didn't solve it at last, I've tried to avoid using scipy in the code, will pick it up again in future if it's a must.
Anyway, thanks for help on this, it can be closed.

**Comment by misl6** [comment]
Fixed via #2615. Feel free to re-open if the issue still persists.
